feat(api): add support for GLM-4.6 model

sarices opened the pull request:
- Adds configuration and normalization for the new GLM-4.6 model.
- Implements `X-Signature` header generation (SHA-256 of request body) to fix upstream `426` errors.
- Updates browser headers (`X-FE-Version`, `User-Agent`) to align with recent frontend changes.
- Includes comprehensive documentation and usage examples for GLM-4.6 in README.md.

sarices committed:
feat(api): add support for GLM-4.6 model

- Adds configuration and normalization for the new GLM-4.6 model.
- Implements `X-Signature` header generation (SHA-256 of request body) to fix upstream `426` errors.
- Updates browser headers (`X-FE-Version`, `User-Agent`) to align with recent frontend changes.
- Includes comprehensive documentation and usage examples for GLM-4.6 in README.md.

sarices committed:
feat(docker): add docker deployment support

This commit introduces a complete Docker-based deployment workflow.

- Adds a multi-stage `Dockerfile` to create an optimized and secure Deno application image.
- Adds a `docker-compose.yml` file for easy one-command startup of the service and any future related services.
- Updates `README.md` and `CLAUDE.md` to include clear instructions for building and running the application using Docker Compose.
- Adds a `.gitignore` file to exclude common files like `.DS_Store` and environment files from version control.

wtklrs commented:
```bash
curl --connect-timeout 0 --max-time 0 --retry 0 http://127.0.0.1:9090/v1/chat/completions \
    -H "Content-Type: application/json" \
    -H "Authorization: Bearer sk-your-key" \
    -d '{
  "model": "GLM-4.5",
  "messages": [
  {
      "role": "system",
      "content": "You are a friendly chatbot who always responds in the style of a pirate"
  },
  {
      "role": "user",
      "content": "How many helicopters can a human eat in one sitting?"
  }
  ],
  "stream": false
}'
```


```bash
$ bash setup.sh
```


```log
OpenAI兼容API服务器启动
支持的模型: 0727-360B-API (GLM-4.5), glm-4.5v (GLM-4.5V), glm-4.6 (GLM-4.6)
上游: https://chat.z.ai/api/chat/completions
Debug模式: true
默认流式响应: true
Dashboard启用: true
运行在本地环境中，端口: 9090
Dashboard已启用，访问地址: http://localhost:9090/dashboard
[DEBUG] 收到chat completions请求
[DEBUG] 🌐 User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.77 Safari/537.36
[DEBUG] API key验证通过
[DEBUG] 📥 收到请求体长度: 290 字符
[DEBUG] 📄 请求体预览: {
"model": "GLM-4.5",
"messages": [
{
"role": "system",
"content": "You are a friendly chatbot who always responds in the style of a pirate"
},
{
"role": "user",
"content": "How many helicopters can a human eat in one sitting?"
}
],
"stream": false
}
[DEBUG] ✅ JSON解析成功
[DEBUG] 🔄 模型ID映射: GLM-4.5 → 0727-360B-API
[DEBUG] 请求解析成功 - 模型: GLM-4.5 (GLM-4.5), 流式: %v, 消息数: NaN 2
[DEBUG] 🔍 Cherry Studio 调试 - 检查原始消息:
[DEBUG]   消息[0] role: system
[DEBUG]   消息[0] content: 字符串类型, 长度: 71
[DEBUG]   消息[0] 内容预览: You are a friendly chatbot who always responds in the style of a pirate
[DEBUG]   消息[1] role: user
[DEBUG]   消息[1] content: 字符串类型, 长度: 52
[DEBUG]   消息[1] 内容预览: How many helicopters can a human eat in one sitting?
[DEBUG] 📝 纯文本消息，长度: 71
[DEBUG] 📝 纯文本消息，长度: 52
[DEBUG] 消息处理完成，处理后消息数: 2
[DEBUG] 匿名token获取成功: eyJhbGciOi...
[DEBUG] 开始处理非流式响应 (chat_id=1759750128033-1759750128)
[DEBUG] 调用上游API: https://chat.z.ai/api/chat/completions
[DEBUG] 上游请求体: {"stream":true,"chat_id":"1759750128033-1759750128","id":"1759750128033","model":"0727-360B-API","messages":[{"role":"system","content":"You are a friendly chatbot who always responds in the style of a pirate"},{"role":"user","content":"Howmany helicopters can a human eat in one sitting?"}],"params":{"top_p":0.95,"temperature":0.6,"max_tokens":80000},"features":{"enable_thinking":true,"image_generation":false,"web_search":false,"auto_web_search":false,"preview_mode":false},"background_tasks":{"title_generation":false,"tags_generation":false},"mcp_servers":[],"model_item":{"id":"0727-360B-API","name":"GLM-4.5","owned_by":"openai","openai":{"id":"0727-360B-API","name":"0727-360B-API","owned_by":"openai","openai":{"id":"0727-360B-API"},"urlIdx":1},"urlIdx":1,"info":{"id":"0727-360B-API","user_id":"api-user","base_model_id":null,"name":"GLM-4.5","params":{"top_p":0.95,"temperature":0.6,"max_tokens":80000},"meta":{"profile_image_url":"/static/favicon.png","description":"Most advanced model, proficient in coding and tool use","capabilities":{"vision":false,"citations":false,"preview_mode":false,"web_search":false,"language_detection":false,"restore_n_source":false,"mcp":true,"file_qa":true,"returnFc":true,"returnThink":true,"think":true}}}},"tool_servers":[],"variables":{"{{USER_NAME}}":"Guest-1759750128033","{{USER_LOCATION}}":"Unknown","{{CURRENT_DATETIME}}":"2025/10/6 19:28:48","{{CURRENT_DATE}}":"2025/10/6","{{CURRENT_TIME}}":"19:28:48","{{CURRENT_WEEKDAY}}":"星期一","{{CURRENT_TIMEZONE}}":"Asia/Shanghai","{{USER_LANGUAGE}}":"zh-CN"}}
[DEBUG] 生成签名: ed684e46300ba4d10f9d46c0ff7e27cdfeb5da592a1929117a3def4c8ce1d1c6 (基于请求体SHA256)
[DEBUG] 上游响应状态: 400 Bad Request
[DEBUG] 上游返回错误状态: 400
```

sarices commented:
> ```shell
> curl --connect-timeout 0 --max-time 0 --retry 0 http://127.0.0.1:9090/v1/chat/completions \
>     -H "Content-Type: application/json" \
>     -H "Authorization: Bearer sk-your-key" \
>     -d '{
>   "model": "GLM-4.5",
>   "messages": [
>   {
>       "role": "system",
>       "content": "You are a friendly chatbot who always responds in the style of a pirate"
>   },
>   {
>       "role": "user",
>       "content": "How many helicopters can a human eat in one sitting?"
>   }
>   ],
>   "stream": false
> }'
> ```
> 
> ```shell
> $ bash setup.sh
> ```
> 
> ```
> OpenAI兼容API服务器启动
> 支持的模型: 0727-360B-API (GLM-4.5), glm-4.5v (GLM-4.5V), glm-4.6 (GLM-4.6)
> 上游: https://chat.z.ai/api/chat/completions
> Debug模式: true
> 默认流式响应: true
> Dashboard启用: true
> 运行在本地环境中，端口: 9090
> Dashboard已启用，访问地址: http://localhost:9090/dashboard
> [DEBUG] 收到chat completions请求
> [DEBUG] 🌐 User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.77 Safari/537.36
> [DEBUG] API key验证通过
> [DEBUG] 📥 收到请求体长度: 290 字符
> [DEBUG] 📄 请求体预览: {
> "model": "GLM-4.5",
> "messages": [
> {
> "role": "system",
> "content": "You are a friendly chatbot who always responds in the style of a pirate"
> },
> {
> "role": "user",
> "content": "How many helicopters can a human eat in one sitting?"
> }
> ],
> "stream": false
> }
> [DEBUG] ✅ JSON解析成功
> [DEBUG] 🔄 模型ID映射: GLM-4.5 → 0727-360B-API
> [DEBUG] 请求解析成功 - 模型: GLM-4.5 (GLM-4.5), 流式: %v, 消息数: NaN 2
> [DEBUG] 🔍 Cherry Studio 调试 - 检查原始消息:
> [DEBUG]   消息[0] role: system
> [DEBUG]   消息[0] content: 字符串类型, 长度: 71
> [DEBUG]   消息[0] 内容预览: You are a friendly chatbot who always responds in the style of a pirate
> [DEBUG]   消息[1] role: user
> [DEBUG]   消息[1] content: 字符串类型, 长度: 52
> [DEBUG]   消息[1] 内容预览: How many helicopters can a human eat in one sitting?
> [DEBUG] 📝 纯文本消息，长度: 71
> [DEBUG] 📝 纯文本消息，长度: 52
> [DEBUG] 消息处理完成，处理后消息数: 2
> [DEBUG] 匿名token获取成功: eyJhbGciOi...
> [DEBUG] 开始处理非流式响应 (chat_id=1759750128033-1759750128)
> [DEBUG] 调用上游API: https://chat.z.ai/api/chat/completions
> [DEBUG] 上游请求体: {"stream":true,"chat_id":"1759750128033-1759750128","id":"1759750128033","model":"0727-360B-API","messages":[{"role":"system","content":"You are a friendly chatbot who always responds in the style of a pirate"},{"role":"user","content":"Howmany helicopters can a human eat in one sitting?"}],"params":{"top_p":0.95,"temperature":0.6,"max_tokens":80000},"features":{"enable_thinking":true,"image_generation":false,"web_search":false,"auto_web_search":false,"preview_mode":false},"background_tasks":{"title_generation":false,"tags_generation":false},"mcp_servers":[],"model_item":{"id":"0727-360B-API","name":"GLM-4.5","owned_by":"openai","openai":{"id":"0727-360B-API","name":"0727-360B-API","owned_by":"openai","openai":{"id":"0727-360B-API"},"urlIdx":1},"urlIdx":1,"info":{"id":"0727-360B-API","user_id":"api-user","base_model_id":null,"name":"GLM-4.5","params":{"top_p":0.95,"temperature":0.6,"max_tokens":80000},"meta":{"profile_image_url":"/static/favicon.png","description":"Most advanced model, proficient in coding and tool use","capabilities":{"vision":false,"citations":false,"preview_mode":false,"web_search":false,"language_detection":false,"restore_n_source":false,"mcp":true,"file_qa":true,"returnFc":true,"returnThink":true,"think":true}}}},"tool_servers":[],"variables":{"{{USER_NAME}}":"Guest-1759750128033","{{USER_LOCATION}}":"Unknown","{{CURRENT_DATETIME}}":"2025/10/6 19:28:48","{{CURRENT_DATE}}":"2025/10/6","{{CURRENT_TIME}}":"19:28:48","{{CURRENT_WEEKDAY}}":"星期一","{{CURRENT_TIMEZONE}}":"Asia/Shanghai","{{USER_LANGUAGE}}":"zh-CN"}}
> [DEBUG] 生成签名: ed684e46300ba4d10f9d46c0ff7e27cdfeb5da592a1929117a3def4c8ce1d1c6 (基于请求体SHA256)
> [DEBUG] 上游响应状态: 400 Bad Request
> [DEBUG] 上游返回错误状态: 400
> ```



昨天还是可用的，今天又不行了，要看看什么原因

notlousybook commented:
@sarices i have the same problems, i have made my own version but i'm unable to figure out why it's not working currently https://github.com/LousyBook94/ZtoApi-Deno

sarices committed:
feat(api): implement dual-layer HMAC-SHA256 signature mechanism

- Add `generateSignature` function to create dual-layer HMAC-SHA256 signatures compliant with Z.ai security specifications.
- Refactor `callUpstreamWithHeaders` to integrate the new signature logic, appending the signature, timestamp, and requestId to upstream requests.
- Remove the old, simplistic signature method based on request body SHA-256.

fix(deno): resolve Deno runtime compatibility issues

- Correct the import for `decode` to `decodeBase64` from `deno/encoding/base64` to align with Deno standard library API changes.
- Fix unused `request` parameter warnings in `handleDashboardStats` and `handleDashboardRequests` functions.

docs: add technical design document for signature mechanism

sarices commented:
> @sarices i have the same problems, i have made my own version but i'm unable to figure out why it's not working currently https://github.com/LousyBook94/ZtoApi-Deno@sarices我有同样的问题，我已经制作了自己的版本，但我无法弄清楚为什么它目前不起作用 https://github.com/LousyBook94/ZtoApi-Deno

已经解决了 https://github.com/roseforyou/ZtoApi/pull/3/commits/c807d30a84e2ffa5f3e38ec9a4505f0f048873dc

notlousybook commented:
@sarices I'm very impressed, i'm wondering how you would figure out the SHA-256 hash, how and where it will be taken from?

wtklrs commented:
不能用了吗? 

在本地运行后显示： `[DEBUG] 响应状态: 405 Not Allowed`

```log
[DEBUG] 响应状态: 405 Not Allowed
[DEBUG] 上游返回错误 body: <!doctypehtml><html lang="zh-cn"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible"content="IE=edge,chrome=1"><meta name="data-spm"content="a3c0e"><title>405</title><style>a,body,div,h2,html,p{margin:0;padding:0}a{text-decoration:none;color:#3b6ea3}.container{width:1000px;margin:auto;color:#696969}.header{padding:110px 0}.header .message{height:36px;padding-left:120px;background:url(https://errors.aliyun.com/images/TB1TpamHpXXXXaJXXXXeB7nYVXX-104-162.png) no-repeat 0 -128px;line-height:36px}.main{padding:50px 0;background:#f4f5f7}#block_image{position:relative;left:120px}</style><body data-spm="7663354"><div data-spm="1998410538"><div class="header"><div class="container"><div class="message"><div id="block_message"></div><div><span id="block_url_tips"></span><strong id="url"></strong></div><div><span id="block_time_tips"></span><strong id="time"></strong></div><div><span id="block_traceid_tips"></span><strong id="traceid"></strong></div></div></div></div><div class="main"><div class="container"><img id="block_image"></div></div></div><script>function getRenderData(){var e=document.getElementById("renderData");return JSON.parse(e.innerHTML)}function convertTimestampToString(e){e=parseInt(e,10),e=new Date(e);return e.getFullYear()+"-"+("0"+(e.getMonth()+1)).slice(-2)+"-"+("0"+e.getDate()).slice(-2)+" "+("0"+e.getHours()).slice(-2)+":"+("0"+e.getMinutes()).slice(-2)+":"+("0"+e.getSeconds()).slice(-2)}var en_tips={block_message:"Sorry, your request has been blocked as it may cause potential threats to the server's security.",block_url_tips:"Current URL: ",block_time_tips:"Request Time: ",block_traceid_tips:"Your Request ID is: "},cn_tips={block_message:"很抱歉，由于您访问的URL有可能对网站造成安全威胁，您的访问被阻断。",block_url_tips:"当前网址: ",block_time_tips:"请求时间: ",block_traceid_tips:"您的请求ID是: "};window.onload=function(){var t=getRenderData(),n="cn";try{navigator.language.startsWith("zh")||(n="en")}catch(e){t.lang&&(n=t.lang)}if(t){var e,i=cn_tips,r=document.getElementById("block_image");for(e in"en"===n?(i=en_tips,r.src="https://g.alicdn.com/sd-base/static/1.0.5/image/405.png",r.id="en_block"):r.src="https://errors.aliyun.com/images/TB15QGaHpXXXXXOaXXXXia39XXX-660-117.png",i)document.getElementById(e).innerText=i[e];n=t.traceid,r=n.slice(8,21);document.getElementById("traceid").innerText=n,document.getElementById("url").innerText=location.href.split("?")[0],document.getElementById("time").innerText=convertTimestampToString(r)}}</script><textarea id="renderData" style="display:none">{"traceid":"0a094eg7176048727688749k3e51b8","lang":"en"}</textarea>
```

sarices commented:
> 不能用了吗?
> 
> 在本地运行后显示： `[DEBUG] 响应状态: 405 Not Allowed`
> 
> ```
> [DEBUG] 响应状态: 405 Not Allowed
> [DEBUG] 上游返回错误 body: <!doctypehtml><html lang="zh-cn"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible"content="IE=edge,chrome=1"><meta name="data-spm"content="a3c0e"><title>405</title><style>a,body,div,h2,html,p{margin:0;padding:0}a{text-decoration:none;color:#3b6ea3}.container{width:1000px;margin:auto;color:#696969}.header{padding:110px 0}.header .message{height:36px;padding-left:120px;background:url(https://errors.aliyun.com/images/TB1TpamHpXXXXaJXXXXeB7nYVXX-104-162.png) no-repeat 0 -128px;line-height:36px}.main{padding:50px 0;background:#f4f5f7}#block_image{position:relative;left:120px}</style><body data-spm="7663354"><div data-spm="1998410538"><div class="header"><div class="container"><div class="message"><div id="block_message"></div><div><span id="block_url_tips"></span><strong id="url"></strong></div><div><span id="block_time_tips"></span><strong id="time"></strong></div><div><span id="block_traceid_tips"></span><strong id="traceid"></strong></div></div></div></div><div class="main"><div class="container"><img id="block_image"></div></div></div><script>function getRenderData(){var e=document.getElementById("renderData");return JSON.parse(e.innerHTML)}function convertTimestampToString(e){e=parseInt(e,10),e=new Date(e);return e.getFullYear()+"-"+("0"+(e.getMonth()+1)).slice(-2)+"-"+("0"+e.getDate()).slice(-2)+" "+("0"+e.getHours()).slice(-2)+":"+("0"+e.getMinutes()).slice(-2)+":"+("0"+e.getSeconds()).slice(-2)}var en_tips={block_message:"Sorry, your request has been blocked as it may cause potential threats to the server's security.",block_url_tips:"Current URL: ",block_time_tips:"Request Time: ",block_traceid_tips:"Your Request ID is: "},cn_tips={block_message:"很抱歉，由于您访问的URL有可能对网站造成安全威胁，您的访问被阻断。",block_url_tips:"当前网址: ",block_time_tips:"请求时间: ",block_traceid_tips:"您的请求ID是: "};window.onload=function(){var t=getRenderData(),n="cn";try{navigator.language.startsWith("zh")||(n="en")}catch(e){t.lang&&(n=t.lang)}if(t){var e,i=cn_tips,r=document.getElementById("block_image");for(e in"en"===n?(i=en_tips,r.src="https://g.alicdn.com/sd-base/static/1.0.5/image/405.png",r.id="en_block"):r.src="https://errors.aliyun.com/images/TB15QGaHpXXXXXOaXXXXia39XXX-660-117.png",i)document.getElementById(e).innerText=i[e];n=t.traceid,r=n.slice(8,21);document.getElementById("traceid").innerText=n,document.getElementById("url").innerText=location.href.split("?")[0],document.getElementById("time").innerText=convertTimestampToString(r)}}</script><textarea id="renderData" style="display:none">{"traceid":"0a094eg7176048727688749k3e51b8","lang":"en"}</textarea>
> ```

"很抱歉，由于您访问的URL有可能对网站造成安全威胁，您的访问被阻断。" 触发敏感链接了
